otaclient/app/boot_control/_rpi_boot.py

```diff
@@ -0,0 +1,521 @@
+)
+from .configs import rpi_boot_cfg as cfg
+from .protocol import BootControllerProtocol
+
+logger = log_setting.get_logger(
+    __name__, cfg.LOG_LEVEL_TABLE.get(__name__, cfg.DEFAULT_LOG_LEVEL)
+)
+
+_FSTAB_TEMPLATE_STR = (
+    "LABEL=${rootfs_fslabel}\t/\text4\tdiscard,x-systemd.growfs\t0\t1\n"
+    "LABEL=system-boot\t/boot/firmware\tvfat\tdefaults\t0\t1\n"
+)
+
+
+class _RPIBootControl:
+    """Boot control helper for rpi4 support.
+
+    Expected partition layout:
+        /dev/sda:
+            - sda1: fat32, fslabel=systemb-boot
+            - sda2: ext4, fslabel=slot_a
+            - sda3: ext4, fslabel=slot_b
+    slot is the fslabel for each AB rootfs.
+
+    This class provides the following features:
+    1. AB partition detection,
+    2. boot files checking,
+    3. switch boot(tryboot.txt setup and tryboot reboot),
+    4. finalize switching boot.
+
+    Boot files for each slot have the following naming format:
+        <boot_file_fname>_<slot>
+    i.e., config.txt for slot_a will be config.txt_slot_a
+    """
+
+    SLOT_A = cfg.SLOT_A_FSLABEL
+    SLOT_B = cfg.SLOT_B_FSLABEL
+    AB_FLIPS = {
+        SLOT_A: SLOT_B,
+        SLOT_B: SLOT_A,
+    }
+    SEP_CHAR = "_"
+
+    def __init__(self) -> None:
+        self.system_boot_path = Path(cfg.SYSTEM_BOOT_MOUNT_POINT)
+        if not (
+            self.system_boot_path.is_dir()
+            and CMDHelperFuncs.is_target_mounted(self.system_boot_path)
+        ):
+            _err_msg = "system-boot is not presented or not mounted!"
+            logger.error(_err_msg)
+            raise BootControlInitError(_err_msg)
+        self._init_slots_info()
+        self._init_boot_files()
+
+    def _init_slots_info(self):
+        """Get current/standby slots info."""
+        logger.debug("checking and initializing slots info...")
+        try:
+            # detect active slot
+            self._active_slot_dev = CMDHelperFuncs.get_dev_by_mount_point(
+                cfg.ACTIVE_ROOTFS_PATH
+            )
+            if not (
+                _active_slot := CMDHelperFuncs.get_fslabel_by_dev(self._active_slot_dev)
+            ):
+                raise ValueError(
+                    f"failed to get slot_id(fslabel) for active slot dev({self._active_slot_dev})"
+                )
+            self._active_slot = _active_slot
+
+            # detect standby slot
+            # NOTE: using the similar logic like grub, detect the silibing dev
+            #       of the active slot as standby slot
+            _parent = CMDHelperFuncs.get_parent_dev(self._active_slot_dev)
+            # list children device file from parent device
+            # exclude parent dev(always in the front)
+            # expected raw result from lsblk:
+            # NAME="/dev/sdx"
+            # NAME="/dev/sdx1" # system-boot
+            # NAME="/dev/sdx2" # slot_a
+            # NAME="/dev/sdx3" # slot_b
+            _raw_child_partitions = CMDHelperFuncs._lsblk(f"-Pp -o NAME {_parent}")
```

Comment:
This PR is OK, but I recently found out that lsblk -J returns JSON string, so it can be used.